Source/Core/VideoCommon/TextureConverterShaderGen.cpp

```diff
@@ -138,147 +153,95 @@ ShaderCode GenerateTextureConverterShaderCode(APIType api_type,
           color_mask[4] = color_mask[5] = color_mask[6] = color_mask[7] = 1.0f / 15.0f;
         }
       }
+      out.Write("	const vec4 colmat[7] = {\n");
+      for (size_t i = 0; i < colmat.size() / 4; i++)
+      {
+        out.Write("		vec4(%f, %f, %f, %f)%s\n", colmat[i * 4 + 0], colmat[i * 4 + 1],
+                  colmat[i * 4 + 2], colmat[i * 4 + 3], i < 7 ? "," : "");
+      }
+      out.Write(
+          "	};\n"
+          "	texcol = floor(texcol * colmat[5]) * colmat[6];\n"
+          "	ocol0 = texcol * mat4(colmat[0], colmat[1], colmat[2], colmat[3]) + colmat[4];\n");
       break;
 
     default:
       ERROR_LOG(VIDEO, "Unknown copy intensity format: 0x%X",
                 static_cast<int>(uid_data->dst_format));
-      colmat[0] = colmat[5] = colmat[10] = colmat[15] = 1.0f;
+      out.Write("	ocol0 = texcol;\n");
       break;
     }
   }
   else
   {
+    if (!uid_data->efb_has_alpha)
+      out.Write("	texcol.a = 1.0;\n");
+
     switch (uid_data->dst_format)
     {
     case EFBCopyFormat::R4:  // R4
-      colmat[0] = colmat[4] = colmat[8] = colmat[12] = 1;
-      color_mask[0] = 255.0f / 16.0f;
-      color_mask[4] = 1.0f / 15.0f;
+      out.Write("	float red = float(int(texcol.r * 255.0) & 0xF0) * (1.0 / 255.0);\n"
```

Comment:
After rethinking about it, I fear you might need:
 * (1.0 / 240.0)
instead of
 * (1.0 / 255.0)